src/rgw/driver/posix/rgw_sal_posix.cc

```diff
@@ -0,0 +1,3090 @@
+    ssize_t ret;
+
+    ret = lseek(obj_fd, ofs, SEEK_SET);
+    if (ret < 0) {
+      ret = errno;
+      ldpp_dout(dpp, 0) << "ERROR: could not seek object " << get_name() << " to "
+	<< ofs << " :" << cpp_strerror(ret) << dendl;
+      return -ret;
+    }
+
+    ret = ::read(obj_fd, read_buf, len);
+    if (ret < 0) {
+      ret = errno;
+      ldpp_dout(dpp, 0) << "ERROR: could not read object " << get_name() << ": "
+	<< cpp_strerror(ret) << dendl;
+      return -ret;
+    }
+
+    bl.append(read_buf, ret);
+
+    return ret;
+  }
+
+  // It's a multipart object, find the correct file, open it, and read it
+  std::string pname;
+  for (auto part : parts) {
+    if (ofs < part.second) {
+      pname = part.first;
+      break;
+    }
+
+    ofs -= part.second;
+  }
+
+  if (pname.empty()) {
+    // ofs is past the end
+    return 0;
+  }
+
+  POSIXObject* shadow_obj;
+  std::unique_ptr<rgw::sal::Object> obj = shadow->get_object(rgw_obj_key(pname));
+  shadow_obj = static_cast<POSIXObject*>(obj.get());
+  int ret = shadow_obj->open(dpp, false);
+  if (ret < 0) {
+    return ret;
+  }
+
+  return shadow_obj->read(ofs, left, bl, dpp, y);
+}
+
+int POSIXObject::write(int64_t ofs, bufferlist& bl, const DoutPrefixProvider* dpp,
+		       optional_yield y)
+{
+  if (shadow) {
+    // Can't write to a MP file
+    return -EINVAL;
+  }
+
+  int64_t left = bl.length();
+  char* curp = bl.c_str();
+  ssize_t ret;
+
+  ret = fchmod(obj_fd, S_IRUSR|S_IWUSR);
+  if(ret < 0) {
+    ldpp_dout(dpp, 0) << "ERROR: could not change permissions on object " << get_name() << ": "
+                  << cpp_strerror(ret) << dendl;
+    return ret;
+  }
+
+
+  ret = lseek(obj_fd, ofs, SEEK_SET);
+  if (ret < 0) {
+    ret = errno;
+    ldpp_dout(dpp, 0) << "ERROR: could not seek object " << get_name() << " to "
+      << ofs << " :" << cpp_strerror(ret) << dendl;
+    return -ret;
+  }
+
+  while (left > 0) {
+    ret = ::write(obj_fd, curp, left);
+    if (ret < 0) {
+      ret = errno;
+      ldpp_dout(dpp, 0) << "ERROR: could not write object " << get_name() << ": "
+	<< cpp_strerror(ret) << dendl;
+      return -ret;
+    }
+
+    curp += ret;
+    left -= ret;
+  }
+
+  return 0;
+}
+
+int POSIXObject::write_attr(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, bufferlist& value)
+{
+  int ret;
+  std::string attrname;
+
+  ret = open(dpp, true);
+  if (ret < 0) {
+    return ret;
+  }
+
+  return write_x_attr(dpp, y, obj_fd, key, value, get_name());
+}
+
+int POSIXObject::POSIXReadOp::prepare(optional_yield y, const DoutPrefixProvider* dpp)
+{
+  int ret = source->stat(dpp);
+  if (ret < 0)
+    return ret;
+
+  ret = source->get_obj_attrs(y, dpp);
+  if (ret < 0)
+    return ret;
+
+  bufferlist etag_bl;
+  if (!rgw::sal::get_attr(source->get_attrs(), RGW_ATTR_ETAG, etag_bl)) {
+    /* Sideloaded file.  Generate necessary attributes. Only done once. */
+    int ret = source->generate_attrs(dpp, y);
```

Comment:
etag seems to be will defined as the md5sum of the contents of the file, so no.  The plan is to eventually add a background service based on fnotify that catches adds and generates etag.

Reply:
this issue doesn't need to block merge, but i'd really like to see further discussion here. this feels like a major scalability issue when compared to other backends, and will be very visible to s3 clients as request timeouts and unpredictable latency

> etag seems to be will defined as the md5sum of the contents of the file, so no

it is defined, but with special cases for multipart and encrypted objects. https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html#checking-object-integrity-etag-and-md5 is a good resource on this

we should consider defining a new special case for the posix backend to avoid this requirement to read a potentially-unbounded amount of data

some s3 clients like s3cmd try to use the etag for checksum verification, but it should generally be possible to disable that with options like `--no-check-md5`. there was a related discussion in https://github.com/s3tools/s3cmd/issues/880, where it looks like minio worked around a similar issue by adding a `-1` suffix so the etag would be interpreted as multipart

aside from the s3 protocol, http has its own requirements for [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) related to its use in caching and the `If-Match`/`If-None-Match` headers. there i believe the only requirement is that the etag changes if the data changes, so some combination of the inode number and mtime may work as a proxy for the md5sum

> The plan is to eventually add a background service based on fnotify that catches adds and generates etag.

i can see how this would reduce the probability that a client reads an object with uninitialized etag so needs to calculate the md5sum inline. but it still imposes a lot of background i/o that shouldn't be necessary

i guess that would also have to cover filesystem writes to existing objects that already have an etag? what if there's an append-only workload, like a log file? this background worker would have to re-read the entire file for checksumming often, yet its actual etag would rarely be consistent

Reply:
I don't have a strong opinion about this.  I added it so that s3cmd would succeed.  It doesn't matter for passing things like s3-tests, since those will never deal with side-loaded data.   If things other than s3cmd really don't check etag in general, then I'm fine with just removing it, and never having an etag set.  The way it's written now is certainly a bandaid, and can't handle may cases, including overwriting or appending outside of RGW, so it can't be a long-term solution.   My impression was that etag was important, and everything used it to verify on every get, but I've never verified that.